### PR TITLE
Support checking the Requires-Python metadata

### LINKED ISSUE #82: Check `python_requires` for supported python version (PEP 440)
EDIT: Running pyroma v4.0 with default pyroma settings on a package described below using setuptools, pyroma reports no python version was specified.

It expects to find a python version in the classifiers field (setup(..., classifiers=[SOME PYTHON VERSION(S) HERE])).

There is 2 places to define supported python versions and for now pyroma only support the classifiers one. 

```
setup(
    ....
    python_requires=">=3.8",
    ....
)
```

### PR DESCRIPTION
**Pull request by davidandreoletti**
Fixes #82

### CODE REVIEW

**Comment by davidandreoletti** [comment]
> First-time contributors need a maintainer to approve running workflows
@CAM-Gerlach  / @regebro  your approval is needed to run CI jobs.

**Comment by CAM-Gerlach** [comment]
I ran them; also, for WIP PRs, you can use GitHub's built-in "Draft" feature rather than manually adding a "WiP" to the title. Review in work now.

File: pyroma/testdata/complete/setup.py
Comment (CAM-Gerlach) [inline pyroma/testdata/complete/setup.py:13]: For completeness, why not have a more complex (and realistic) specifierset here (and the others)?

```suggestion
    python_requires=">=2.6,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*",
```

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
I confess that I don't really understand the motivation for gloming this functionality on to the existing classifier check when this is really an entirely separate, independent check. Conflated them, as well as implementing rather complex handling of various interactions between them, just complicates the logic and implementation of the existing check and furthermore makes what is already the longest and most complicated check even longer and more complicated. This adds technical risk to this PR, making reviewing and testing it more difficult and increases future maintenance burden, for little clear gain.

Instead of doing all this, I would instead advise just making it a separate check, and simply tweaking the message of the existing check to be specific to the version classifiers. I.e., as I suggested on the issue, just

```python
class PythonVersionRequires(BaseTest):
    weight = 100

    def test(self, data):
        requires_python = data.get("requires_python")
        if not requires_python:
            return False
        try:
            packaging.specifiers.SpecifierSet(requires_python)
        except packaging.specifiers.InvalidSpecifier:
            return False
        return True

    def message(self):
        return (
            "You should specify what Python versions you support "
            "with a valid version specifier in the "
            "requires-python/python_requires metadata key"
        )
```

File: pyroma/testdata/complete/setup.py
Comment (davidandreoletti) [reply pyroma/testdata/complete/setup.py:13]: > I don't really understand the motivation for gloming this functionality on to the existing classifier test when this is really an entirely separate, independent check. 

I initially started it as an independent test like in your snippet. However, at least as I understood it reading the code, pyroma used `classfiers ` for python version checking as the **reference**, ie if you had a "python version classifier" then your lib was "well defined" as a python package.

If it was up to me, I would have moved the _classifier python version check_ to _advisory only_ (ie the test would not contribute to the '10 points' rating) and used only the _python-requires check_ for _rating_ purpose because that's what pip, poetry and others tools use.

My opinion is irrelevant though: pyroma users have been using classifiers this way. 

**I do not wish to break package rating** without first informing users. That's the maintainer's job to decide how / when to do this + communicate this to end users (deprecation policy, etc).

This decision has not been made. As such, I did not see how to both checks independently without breaking existing behaviour.

> This adds technical risk to this PR, making reviewing and testing it more difficult and increases future maintenance burden, for little clear gain.

The existing code has been improved (clearer return check / weight ).

> just complicates the logic and implementation of the existing check and furthermore makes what is already the longest and most complicated check even longer and more complicated

Well, opinion differs here. Hopefully, the first 3 paragraph clarify my intent.

**Willing to discuss this further and revert to your preferred solution.**

Comment (davidandreoletti) [reply pyroma/testdata/complete/setup.py:13]: > For completeness, why not have a more complex (and realistic) specifierset here (and the others)?

Testing a simple case is sufficient to illustrate pyroma perform the right check. More complex case would not run additional pyroma lines of code (but will in packaging.specifiers which is tasked to verify the version requirements).

Comment (davidandreoletti) [reply pyroma/testdata/complete/setup.py:13]: > MO, I confess that I don't really understand the motivation for gloming this functionality on to the existing classifier test when this is really an entirely separate, independent check. Conflated them, as well as implementing rather complex handling of various interactions between them, just complicates the logic and implementation of the existing check and furthermore makes what is already the longest and most complicated check even longer and more complicated. This adds technical risk to this PR, making reviewing and testing it more difficult and increases future maintenance burden, for little clear gain.

If I do what you suggested, then most test fails with 'You should specify what Python version ... with 'python_requires=...''. 

Not sure you guys want to break user packages in the next release (even though it is the right thing to do).

Comment (davidandreoletti) [reply pyroma/testdata/complete/setup.py:13]: Side note: we are not the only ones to bump into [the classifier for humans , python_requires for machine](https://github.com/badges/shields/issues/5551).

**Comment by regebro** [comment]
It is a good point that they should state the sameish thing.

**Comment by davidandreoletti** [comment]
> this is really an entirely separate, independent check.

You are right. The initial version tried not to break compatibility with current release: package without a "python_requires" are well "package" from pyroma's point of view.

With the new version (independent test added), then most test fails with 'You should specify what Python version ... with 'python_requires=...''. Tests have been fixed but pyroma users will have to add "python_requires" to their package description file starting from the next release.

File: pyroma/testdata/complete/setup.py
Comment (davidandreoletti) [reply pyroma/testdata/complete/setup.py:13]: > > For completeness, why not have a more complex (and realistic) specifierset here (and the others)?
> 
> Testing a simple case is sufficient to illustrate pyroma perform the right check. More complex yet valid case would not run additional pyroma lines of code (but will in `packaging` package which is tasked to verify various version requirements syntax).

Yet, I still added case for good measure ...

File: pyroma/ratings.py
Comment (regebro) [inline pyroma/ratings.py:234]: So this is just changing which quotes to use, which seems very spurious. Anyway, I realized through this that I hadn't enforced `black` so I did that.

Please rebase this branch from master, and black should refuse this change. Sorry I hadn't enforced it earlier.

**Review by regebro — CHANGES_REQUESTED** [review]

File: pyroma/ratings.py
Comment (davidandreoletti) [reply pyroma/ratings.py]: done.

File: pyroma/ratings.py
Comment (CAM-Gerlach) [inline pyroma/ratings.py:286]: What is going on here is that some sources of data for Pyroma are reading the tool-specific source configuration of the build backend (e.g. `setup.cfg` or `setup.py` for Setuptools), which use one of several different keys (e.g. `python_requires` for Setuptools, or `requires-python` for PEP 621) whereas others are reading the output core metadata out of the `METADATA` file, which has the `Requires-Python` field.

However, instead of trying to handle all of that here, simply add a line to the [METADATA_MAP dictionary](https://github.com/regebro/pyroma/blob/606db052ef127e664dd52d59ff57c6828974bfed/pyroma/projectdata.py#L15) mapping `requires_python` to `python_requires` (this project uses the legacy Setuptools-specific keys for historical reasons instead of the standard PEP 621 keys or core metadata fields, as it was around long before PEP 517 or 621 existed), and elide all this code here:

```suggestion
        python_requires = data.get("python_requires", None)
```

File: pyroma/ratings.py
Comment (CAM-Gerlach) [inline pyroma/ratings.py:237]: Seems like there are still spurious formatting changes being injected here

File: pyroma/ratings.py
Comment (CAM-Gerlach) [inline pyroma/ratings.py:272]: And here

File: pyroma/ratings.py
Comment (CAM-Gerlach) [inline pyroma/ratings.py:273]:
```suggestion
        return "The classifiers should specify what Python versions you support."
```

This still needs to be updated.

File: README.rst
Comment (CAM-Gerlach) [inline README.rst:75]: Instead of only using a legacy tool-specific key used only by Setuptools (which itself is ditching it eventually now that it has moved to PEP 621 itself), you should at the very least mention the standard PEP 621 key here. Also, use proper reST syntax.

```suggestion
* You should have ``requires-python``/``python_requires``
  specifying the Python versions you support.
```

File: pyroma/ratings.py
Comment (CAM-Gerlach) [inline pyroma/ratings.py:24]: The `specifiers` submodule needs to be imported explicitly here:

```suggestion
import packaging.specifiers
from docutils.core import publish_parts
from docutils.utils import SystemMessage
from trove_classifiers import classifiers as CLASSIFIERS
```

File: pyroma/ratings.py
Comment (CAM-Gerlach) [inline pyroma/ratings.py:296]: As a general best practice, your `try` block should be as minimal as possible and only the line(s) of code in which the error is actually expected to occur should be included within, otherwise you might swallow errors you don't intend. Also, its not clear why you're importing the necessary submodule down here inside the `try` block, instead of at the top like everything else (whereas `packaging` is already imported at the top level, but never used).

Therefore, you should do (as in my example)

```suggestion
        try:
            packaging.specifiers.SpecifierSet(python_requires)
        except packaging.specifiers.InvalidSpecifier as e:
            return False
        return True
```

File: pyroma/ratings.py
Comment (CAM-Gerlach) [inline pyroma/ratings.py:300]: This is not only Setuptools- but `setup.py`-only syntax, which is now heavily discouraged for most modern projects. Therefore, we should say something like

```suggestion
            "You should specify what Python versions you support with "
            "the 'requires-python'/'python_requires' metadata."
```

File: pyroma/tests.py
Comment (CAM-Gerlach) [inline pyroma/tests.py:128]:
```suggestion
                    (
                        "You should specify what Python versions you support with "
                        "the 'requires-python'/'python_requires' metadata."
                    )
```

Update to match above

File: pyroma/tests.py
Comment (CAM-Gerlach) [inline pyroma/tests.py:164]:
```suggestion
                    (
                        "You should specify what Python versions you support with "
                        "the 'requires-python'/'python_requires' metadata."
                    )
```

File: pyroma/tests.py
Comment (CAM-Gerlach) [inline pyroma/tests.py:189]:
```suggestion
                    (
                        "You should specify what Python versions you support with "
                        "the 'requires-python'/'python_requires' metadata."
                    )
```

File: pyroma/tests.py
Comment (CAM-Gerlach) [inline pyroma/tests.py:217]:
```suggestion
                    (
                        "You should specify what Python versions you support with "
                        "the 'requires-python'/'python_requires' metadata."
                    )
```

File: pyroma/tests.py
Comment (CAM-Gerlach) [inline pyroma/tests.py:163]:
```suggestion
                    "The classifiers should specify what Python versions you support.",
```

This also needs an update to match

File: pyroma/tests.py
Comment (CAM-Gerlach) [inline pyroma/tests.py:188]:
```suggestion
                    "The classifiers should specify what Python versions you support.",
```

File: pyroma/tests.py
Comment (CAM-Gerlach) [inline pyroma/tests.py:216]:
```suggestion
                    "The classifiers should specify what Python versions you support.",
```

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
This is looking a lot better, thanks, though it does need some more work...

File: pyroma/ratings.py
Comment (davidandreoletti) [reply pyroma/ratings.py]: Got it. Thanks.

Comment (davidandreoletti) [reply pyroma/ratings.py]: Thanks for the explanation.

**Comment by davidandreoletti** [comment]
@CAM-Gerlach Updated.

**Review by CAM-Gerlach — COMMENTED** [review]
Sorry for the long delay, was working long hours due to the Artemis 1 mission and then had to rest and catch up after. Looks good now; just a couple relatively minor tweaks and this should be ready to go, thanks.

**Comment by CAM-Gerlach** [comment]
Just FYI, with GitHub's suggestion feature, you can just go to the `Files` tab, press `Add to batch` on the suggestions you want to apply and then commit them with the `Commit` button, which then commits the suggestion the reviewer is requesting and automatically resolves the comment for you. This saves you a lot of work making the change manually, and is also a lot less for reviewers to confirm the requested changes were made and see what other modifications were done.

> Once approved, should I squash the PR's commits or will you merge-squash it?

That's up to @regebro

**Comment by regebro** [comment]
Merge-squashing is fine.

**Review by regebro — APPROVED** [review]

**Pull request merged**
